sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -63,10 +63,16 @@ object PhysicalOperation extends PredicateHelper {
         val substitutedFields = fields.map(substitute(aliases)).asInstanceOf[Seq[NamedExpression]]
         (Some(substitutedFields), filters, other, collectAliases(substitutedFields))
 
-      case Filter(condition, child) if condition.deterministic =>
-        val (fields, filters, other, aliases) = collectProjectsAndFilters(child)
-        val substitutedCondition = substitute(aliases)(condition)
-        (fields, filters ++ splitConjunctivePredicates(substitutedCondition), other, aliases)
+      case filter @ Filter(condition, child)
+          if condition.deterministic || condition.isInstanceOf[And] =>
+        val determinedCondition = getDeterminedExpression(condition)
+        if (determinedCondition.isDefined) {
+          val (fields, filters, other, aliases) = collectProjectsAndFilters(child)
+          val substitutedCondition = substitute(aliases)(determinedCondition.get)
+          (fields, filters ++ splitConjunctivePredicates(substitutedCondition), other, aliases)
+        } else {
+          (None, Nil, filter, Map.empty)
+        }
```

Comment:
Could you add tests somewhere to check if this addition could extract deterministic conditions you expect?

Reply:
I have added the test case named `Partition pruning - with filter containing non-determined condition in sub And-expr` in `PruningSuite`.

Reply:
Yea, but they are end-to-end tests, so I think we need more fine-grained tests for `collectProjectsAndFilters`.